dpnp/dpnp_iface_histograms.py

```diff
@@ -393,48 +454,79 @@ def histogram(a, bins=10, range=None, density=None, weights=None):
 
     a, weights, usm_type = _ravel_check_a_and_weights(a, weights)
 
-    bin_edges, uniform_bins = _get_bin_edges(a, bins, range, usm_type)
+    bin_edges, _ = _get_bin_edges(a, bins, range, usm_type)
 
     # Histogram is an integer or a float array depending on the weights.
     if weights is None:
         ntype = dpnp.dtype(dpnp.intp)
     else:
         ntype = weights.dtype
 
-    # The fast path uses bincount, but that only works for certain types
-    # of weight
-    # simple_weights = (
-    #     weights is None or
-    #     dpnp.can_cast(weights.dtype, dpnp.double) or
-    #     dpnp.can_cast(weights.dtype, complex)
-    # )
-    # TODO: implement a fast path
-    simple_weights = False
-
-    if uniform_bins is not None and simple_weights:
-        # TODO: implement fast algorithm for equal bins
-        pass
-    else:
-        # Compute via cumulative histogram
-        if weights is None:
-            sa = dpnp.sort(a)
-            cum_n = _search_sorted_inclusive(sa, bin_edges)
-        else:
-            zero = dpnp.zeros(
-                1, dtype=ntype, sycl_queue=a.sycl_queue, usm_type=usm_type
-            )
-            sorting_index = dpnp.argsort(a)
-            sa = a[sorting_index]
-            sw = weights[sorting_index]
-            cw = dpnp.concatenate((zero, sw.cumsum(dtype=ntype)))
-            bin_index = _search_sorted_inclusive(sa, bin_edges)
-            cum_n = cw[bin_index]
+    queue = a.sycl_queue
+    device = queue.sycl_device
+
+    a_bin_dtype, hist_dtype = _align_dtypes(
+        a.dtype, bin_edges.dtype, ntype, device
+    )
+
+    if a_bin_dtype is None or hist_dtype is None:
+        raise ValueError(
+            f"function '{histogram}' does not support input types "
+            f"({a.dtype}, {bin_edges.dtype}, {ntype}), "
+            "and the inputs could not be coerced to any "
+            "supported types"
+        )
 
-        n = dpnp.diff(cum_n)
+    a_casted = dpnp.astype(a, a_bin_dtype, order="C", copy=False)
+    bin_edges_casted = dpnp.astype(
+        bin_edges, a_bin_dtype, order="C", copy=False
+    )
+    weights_casted = (
+        dpnp.astype(weights, hist_dtype, order="C", copy=False)
+        if weights is not None
+        else None
+    )
+
+    # histogram implementation uses atomics, but atomics doesn't work with
+    # host usm memory
+    n_usm_type = "device" if usm_type == "host" else usm_type
+
+    n_casted = dpnp.zeros(
```

Comment:
Do we need to fill the memory with zeros?
```suggestion
    n_casted = dpnp.empty(
```

Reply:
We need, yes.
Histogram atomically increments elements of output array. So it must be set to zero in the beginning.

Reply:
It would be helpful to add a comment somehwere clarifying that `statistics_ext.histogram` expects result array with zero values.

Reply:
added comment